File: Convert Sorted Array to Binary Search Tree/Convert Sorted Array to Binary Search Tree.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param num, a list of integers
    # @return a tree node
    def sortedArrayToBST(self, num):
        if len(num)==0:
            return None
        return self.ArrayToBST(num, len(num))
    
    def ArrayToBST(self, num, n):
        if n==1:
            return TreeNode(num[0])
        if n==2:
            root=TreeNode(num[1])
            left=TreeNode(num[0])
            root.left=left
            return root
        else:
            numl=[num[i] for i in range(n/2)]
            numr=[[num[j] for j in range(n/2+1,n)]
            l=self.ArrayToBST(numl, n/2)
            r=self.ArrayToBST(numr, (n-1)/2)
            root=TreeNode(num[n/2])
            root.left=l
            root.right=r
            return root

File: Convert Sorted Array to Binary Search Tree/Convert Sorted Array to Binary Search Tree.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode sortedArrayToBST(int[] num) {
        if(num.length==0) return null;
        return ArrayToBST(num, num.length);
    }
    public TreeNode ArrayToBST(int[] num, int n){
        if(n==1) return new TreeNode(num[0]);
        if(n==2){
            TreeNode root=new TreeNode(num[1]);
            TreeNode left=new TreeNode(num[0]);
            root.left=left;
            return root;
        }
        int[] numl=new int[n/2];
        int[] numr=new int[(n-1)/2];
        for(int i=0;i<n/2;i++){
            numl[i]=num[i];
        }
        for(int j=n/2+1;j<n;j++){
            numr[j-n/2-1]=num[j];
        }
        TreeNode l=ArrayToBST(numl, n/2);
        TreeNode r=ArrayToBST(numr, (n-1)/2);
        TreeNode root=new TreeNode(num[n/2]);
        root.left=l;
        root.right=r;
        return root;
    }
}
